File: src/main/java/com/pedigree/ui/SelectionBus.java
package com.pedigree.ui;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;

/**
 * Simple UI-level selection bus to propagate selected entity id
 * (e.g., from canvas to list views).
 */
public final class SelectionBus {
    private static final List<Consumer<String>> LISTENERS = new CopyOnWriteArrayList<>();

    private SelectionBus() {}

    public static void addListener(Consumer<String> listener) {
        if (listener != null) {
            LISTENERS.add(listener);
        }
    }

    public static void removeListener(Consumer<String> listener) {
        LISTENERS.remove(listener);
    }

    public static void publish(String id) {
        for (Consumer<String> l : LISTENERS) {
            try {
                l.accept(id);
            } catch (Throwable ignored) {
                // avoid breaking the chain
            }
        }
    }
}

File: src/main/kotlin/com/family/tree/ui/SelectionBus.kt
package com.family.tree.ui

import java.util.concurrent.CopyOnWriteArrayList
import java.util.function.Consumer

/**
 * Simple UI-level selection bus to propagate selected entity id
 * (e.g., from canvas to list views).
 */
object SelectionBus {
    private val LISTENERS: MutableList<Consumer<String>> = CopyOnWriteArrayList()

    @JvmStatic
    fun addListener(listener: Consumer<String>?) {
        if (listener != null) {
            LISTENERS.add(listener)
        }
    }

    @JvmStatic
    fun removeListener(listener: Consumer<String>?) {
        if (listener != null) LISTENERS.remove(listener)
    }

    @JvmStatic
    fun publish(id: String) {
        for (l in LISTENERS) {
            try {
                l.accept(id)
            } catch (_: Throwable) {
                // avoid breaking the chain
            }
        }
    }
}
